JS: add query for detecting ignored calls to Array.prototype.concat

erik-krogh opened the pull request:
A simple and precise query for detecting ignored calls to `Array.prototype.concat`.  

These are the current results: https://lgtm.com/query/6200487140988633440/

It started out as a [query for detecting ignored calls to pure functions](https://lgtm.com/query/2784999926000416306/), but that query had too many weird results.

Type-tracking is used to detect if the receiver is an array, otherwise the query would flag various `concat` methods from different libraries.

Type-tracking is similarly used to detect if the argument to `Array.prototype.concat` is an array.  
This might be a bit too restrictive, and I'm considering removing that restriction.  
But it by keeping it the query doesn't flag a log of unit-tests of the `concat` method itself. 
These are the results we remove from restricting the argument to be an array: 
https://lgtm.com/query/1009789502976493454/  
(An example of why I'm considering removing that restriction: [Someone called `concat` instead of `push`.](https://lgtm.com/projects/g/ttezel/twit/snapshot/1fb3375090c0416950b1493381868b673ee239b9/files/lib/twitter.js?sort=name&dir=ASC&mode=heatmap#x92d46ea3c62eb565:1))

[Performance is good.](https://git.semmle.com/erik/dist-compare-reports/tree/augustus.ti.semmle.com_1571994373309)   
There is only a slight outlier with `closure-library`, where the type-tracking of arrays take a while.

asger-semmle commented:
Have you tried backtracking from the receiver of a `concat` instead of forward-tracking all arrays? That might perform better on closure-library.

max-schaefer commented on javascript/ql/src/Statements/IgnoreConcatReturn.ql:
```diff
@@ -0,0 +1,42 @@
+  call.getCalleeName() = "concat" and
+  call.getNumArgument() = 1 and 
+  (call.getArgument(0).getALocalSource() = array() or isIncomplete(call.getArgument(0))) and 
+  not call.getArgument(0).asExpr().(ArrayExpr).getSize() = 0 and 
```
What about the receiver being the empty array?

erik-krogh replied:
The call does nothing if the input is literally the empty array, and I got tired of looking at [tests like these](https://lgtm.com/projects/g/binoc-graveyard/mozilla-esr24/snapshot/787bf918e25e292675d5977c105c825b0bf5fe47/files/js/src/jit-test/tests/auto-regress/bug528048.js?sort=name&dir=ASC&mode=heatmap#xafd0777a79296517:1).

max-schaefer replied:
I'm not sure I understand your answer. I do agree with this whitelisting. I was just wondering about the symmetric case of `[].concat(xs)`, which I saw in a few tests.

erik-krogh replied:
> Is this PR ready for an editorial review or shall I hold off until all the technical discussions have taken place?

Hold off for now.

erik-krogh replied:
Oh. I misread your comment, sorry. I was thinking about the way I wrote the query and thought you were talking about the argument not the receiver. 

You have a good point with the receiver being the empty array. I think we should remove those.

max-schaefer commented on javascript/ql/src/Statements/IgnoreConcatReturn.ql:
```diff
@@ -0,0 +1,42 @@
+from DataFlow::CallNode call
+where 
+  isArrayMethod(call) and
+  call.getCalleeName() = "concat" and
```
I think those two lines can be simplified to

```suggestion
  call = array().getAMethodCall("concat")
```

(Though that won't work if you do type backtracking instead, which I think is a good idea.)

mchammer01 commented:
Is this PR ready for an editorial review or shall I hold off until all the technical discussions have taken place?

erik-krogh commented:
> Is this PR ready for an editorial review or shall I hold off until all the technical discussions have taken place?

Hold off for now.

erik-krogh commented:
I implemented the backtracking instead, and it did indeed fix the performance of `closure-library`.  
However, I have to relate the array-creation with the call to the array method, and I therefore ended up with the `call` as an argument through the type-backtracking.  
Its not pretty, but it works.

We risk a performance-blowup if the `isIgnoredPureArrayCall` predicate is large, but it should be small in practice, so I don't think there is an issue. 

With the backtracking there was not pretty way to restrict the `concat` calls to only the ones that accepted Arrays, so I removed that part of the query. (As mentioned before, I considered removing it anyway). 

I'm doing a new performance evaluation on defaults right now.

erik-krogh commented:
> Taking a step back, would it perhaps make sense to generalise this query to look for any array method that is annotated with @nosideeffects in our externs definitions? I imagine, for instance, that developers might occasionally use array.slice where they meant array.splice (the former, of course, not having any side effects).

I added the `join`/`slice` methods.  
I haven't seen how/if they are misused, so I don't know the precision of the query for those methods. 

The remaining `@nosideeffects` methods are more precisely handled here: https://github.com/Semmle/ql/pull/2116

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.qhelp:
```diff
@@ -0,0 +1,48 @@
+<overview>
+<p>
+The <code>concat</code>, <code>join</code> and <code>slice</code> methods are
+pure and does not modify any of the inputs or the array the method was called 
```
```suggestion
pure and do not modify any of the inputs or the array the method was called 
```

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.ql:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @name Ignoring result from pure array method
+ * @description The array methods do not modify the array, ignoring the result of such a call is therefore generally an error.
```
```suggestion
 * @description Ignoring the result of an array method that does not modify its receiver is generally an error.
```

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.ql:
```diff
@@ -0,0 +1,47 @@
+ * @kind problem
+ * @problem.severity warning
+ * @id js/ignore-array-result
+ * @tags maintainability,
```
```suggestion
 * @tags maintainability
```

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.ql:
```diff
@@ -0,0 +1,47 @@
+import Expressions.ExprHasNoEffect
+
+DataFlow::SourceNode callsArray(DataFlow::TypeBackTracker t, DataFlow::MethodCallNode call) {
+  isIgnoredPureArrayCall(call) and
```
It may be better to distribute this into the first disjunct. It is implied by the recursive call in the second disjunct.

erik-krogh replied:
It definitely makes it more readable, as it now follows the usual type-tracking pattern.

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.ql:
```diff
@@ -0,0 +1,47 @@
+  isIgnoredPureArrayCall(call) and
+  (
+    t.start() and
+    result = call.getReceiver()
```
```suggestion
    result = call.getReceiver().getALocalSource()
```

max-schaefer commented on javascript/ql/src/Statements/IgnoreArrayResult.ql:
```diff
@@ -0,0 +1,47 @@
+from DataFlow::MethodCallNode call
+where 
+  callsArray(call) instanceof DataFlow::ArrayCreationNode and
+  not call.getReceiver().asExpr().(ArrayExpr).getSize() = 0
```
Don't you want to check the size of the result of `callsArray` here?

erik-krogh replied:
No. I want to allow e.g. 

```JavaScript
var foo = [];
addStuffTo(foo);
foo.concat(other);
```

I have found TP's that are missed by just changing to `call.getReceiver().getALocalSource()`.

max-schaefer replied:
Makes sense.

erik-krogh commented:
> Apologies, I should have explained by suggestion about `@nosideeffects` a little more clearly

I did find that that part of the code, and decided not to use it.  
E.g. `toString` is marked as `@nosideeffects`. But when I looked at pure functions in general I found plenty of benign uses of exactly `toString`. So I prefer having an explicit white-list in the query.

max-schaefer commented on change-notes/1.23/analysis-javascript.md:
```diff
@@ -23,6 +23,7 @@
 | Use of returnless function (`js/use-of-returnless-function`)              | maintainability, correctness                                      | Highlights calls where the return value is used, but the callee never returns a value. Results are shown on LGTM by default. |
 | Useless regular expression character escape (`js/useless-regexp-character-escape`) | correctness, security, external/cwe/cwe-20 | Highlights regular expression strings with useless character escapes, indicating a possible violation of [CWE-20](https://cwe.mitre.org/data/definitions/20.html). Results are shown on LGTM by default. |
 | Unreachable method overloads (`js/unreachable-method-overloads`)          | correctness, typescript                                           | Highlights method overloads that are impossible to use from client code. Results are shown on LGTM by default. |
+| Ignoring return from concat (`js/ignore-return-from-concat`)              | maintainability, correctness                                      | Highlights calls to the concat method on array where the return value is ignored. Results are shown on LGTM by default. |
```
Needs updating now that the query is more general.

mchammer01 commented:
Thanks @max-schaefer - I'll review this first thing on Thursday (not in tomorrow).

mchammer01 commented on javascript/ql/src/Statements/IgnoreArrayResult.qhelp:
```diff
@@ -0,0 +1,48 @@
+<overview>
+<p>
+The <code>concat</code>, <code>join</code> and <code>slice</code> methods are
+pure and do not modify any of the inputs or the array the method was called 
```
Just a question here. Are you deliberately using the past here? ….was called...

erik-krogh replied:
no..
I'm changing it to `is called`.

mchammer01 commented on javascript/ql/src/Statements/IgnoreArrayResult.qhelp:
```diff
@@ -0,0 +1,48 @@
+<example>
+
+<p>
+In the following example a function <code>extend</code> is defined. The 
```
```suggestion
A function <code>extend</code> is defined in the following example. The 
```

mchammer01 commented on javascript/ql/src/Statements/IgnoreArrayResult.qhelp:
```diff
@@ -0,0 +1,48 @@
+
+<p>
+Assigning the returned value from the call to <code>concat</code> to the 
+<code>arr</code> variable fixes the error.
```
```suggestion
<code>arr</code> variable fixes the error:
```

mchammer01 commented on javascript/ql/src/Statements/IgnoreArrayResult.qhelp:
```diff
@@ -0,0 +1,48 @@
+In the following example a function <code>extend</code> is defined. The 
+function uses the <code>concat</code> method to add elements to the 
+<code>arr</code> array. However, the <code>extend</code> function has no 
+effect as the return value from <code>concat</code> is ignored.
```
```suggestion
effect as the return value from <code>concat</code> is ignored:
```

mchammer01 commented on change-notes/1.23/analysis-javascript.md:
```diff
@@ -23,6 +23,7 @@
 | Use of returnless function (`js/use-of-returnless-function`)              | maintainability, correctness                                      | Highlights calls where the return value is used, but the callee never returns a value. Results are shown on LGTM by default. |
 | Useless regular expression character escape (`js/useless-regexp-character-escape`) | correctness, security, external/cwe/cwe-20 | Highlights regular expression strings with useless character escapes, indicating a possible violation of [CWE-20](https://cwe.mitre.org/data/definitions/20.html). Results are shown on LGTM by default. |
 | Unreachable method overloads (`js/unreachable-method-overloads`)          | correctness, typescript                                           | Highlights method overloads that are impossible to use from client code. Results are shown on LGTM by default. |
+| Ignoring return from concat (`js/ignore-return-from-concat`)              | maintainability, correctness                                      | Highlights calls to the concat method on array where the return value is ignored. Results are shown on LGTM by default. |
```
Shouldn't the name be `Ignoring result from pure array method`, and shouldn't the description mention the other two methods?

erik-krogh replied:
Correct. I had forgot to change that when the scope of the query changed.

max-schaefer commented:
Almost there! Could you add the query to https://github.com/Semmle/ql/blob/master/javascript/config/suites/javascript/correctness-core, please?

erik-krogh commented:
@mchammer01 a final review?

erik-krogh commented:
I did a final [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/balbinus.ti.semmle.com_1573028379478). Performance and results are still looking good.

mchammer01 commented:
Thanks for the ping @erik-krogh - I have now approved this PR.
